core/org.wso2.carbon.user.core/src/main/java/org/wso2/carbon/user/core/config/UserStoreConfigXMLProcessor.java

```diff
@@ -430,56 +434,53 @@ private static PrivateKey getPrivateKey() {
      * @throws BadPaddingException
      * @throws IllegalBlockSizeException
      */
-    private String decryptProperty(String propValue)
-            throws NoSuchPaddingException, NoSuchAlgorithmException, NoSuchProviderException,
-            org.wso2.carbon.user.api.UserStoreException, InvalidKeyException, BadPaddingException,
-            IllegalBlockSizeException {
+    private String decryptProperty(String propValue) throws CryptoException {
 
-        Cipher keyStoreCipher;
         String cipherTransformation = System.getProperty(CIPHER_TRANSFORMATION_SYSTEM_PROPERTY);
         byte[] cipherTextBytes = Base64.decode(propValue.trim());
+        String algorithm = null;
+        byte[] decryptedValue;
 
-        privateKey = (privateKey == null) ? getPrivateKey() : privateKey;
-        if (privateKey == null) {
-            throw new org.wso2.carbon.user.api.UserStoreException(
-                    "Private key initialization failed. Cannot decrypt the userstore password.");
-        }
+        CryptoService cryptoService = UserStoreMgtDataHolder.getInstance().getCryptoService();
 
-        if(cipherTransformation != null) {
+        if (cipherTransformation != null) {
             // extract the original cipher if custom transformation is used configured in carbon.properties.
-            CipherHolder cipherHolder = cipherTextToCipherHolder(cipherTextBytes);
+            CipherMetaDataHolder cipherHolder = cipherTextToCipherHolder(cipherTextBytes);
             if (cipherHolder != null) {
                 // cipher with meta data.
                 if (log.isDebugEnabled()) {
                     log.debug("Cipher transformation for decryption : " + cipherHolder.getTransformation());
                 }
-                keyStoreCipher = Cipher.getInstance(cipherHolder.getTransformation(), "BC");
+                algorithm = cipherHolder.getTransformation();
                 cipherTextBytes = cipherHolder.getCipherBase64Decoded();
             } else {
                 // If the ciphertext is not a self-contained, directly decrypt using transformation configured in
                 // carbon.properties file
-                keyStoreCipher = Cipher.getInstance(cipherTransformation, "BC");
+                algorithm = cipherTransformation;
+            }
+        }
+        if (cipherTextBytes.length == 0) {
+            decryptedValue = StringUtils.EMPTY.getBytes();
+            if (log.isDebugEnabled()) {
+                log.debug("Ciphertext is empty. An empty array will be used as the plaintext bytes.");
             }
         } else {
-            // If reach here, user have removed org.wso2.CipherTransformation property or carbon.properties file
-            // hence RSA is considered as default transformation
-            keyStoreCipher = Cipher.getInstance("RSA", "BC");
+            decryptedValue = cryptoService.decrypt(cipherTextBytes, algorithm, CRYPTO_API_PROVIDER_BC);
```

Comment:
what will happen if cipherTransformation is null. Algorithm will be null and default on will be resolved from crypto service ?

Reply:
Yes, default one will be resolved from the respective internal crypto provider in the crypto service